reputation/reputation.go

```diff
@@ -19,39 +25,165 @@ type ReputationModule struct {
 	ds      datastore.TxnDatastore
 	sources *source.SourceStore
 
+	mi *miner.MinerIndex
+
+	si *slashing.SlashingIndex
+
+	ai *ask.AskIndex
+
+	lockIndex sync.Mutex
+	mIndex    miner.Index
+	sIndex    slashing.Index
+	aIndex    ask.Index
+
+	lockScores sync.Mutex
+	rebuild    chan struct{}
+	scores     []MinerScore
+
 	ctx    context.Context
 	cancel context.CancelFunc
 }
 
-func New(ds datastore.TxnDatastore) *ReputationModule {
+type MinerScore struct {
+	Addr  string
+	Score int
+}
+
+func New(ds datastore.TxnDatastore, mi *miner.MinerIndex, si *slashing.SlashingIndex, ai *ask.AskIndex) *ReputationModule {
 	ctx, cancel := context.WithCancel(context.Background())
 	rm := &ReputationModule{
 		ds: ds,
+		mi: mi,
+		si: si,
+		ai: ai,
 
+		rebuild: make(chan struct{}, 1),
 		ctx:     ctx,
 		cancel:  cancel,
 		sources: source.NewStore(ds),
 	}
 
 	go rm.updateSources()
+	go rm.subscribeIndexes()
+	go rm.indexBuilder()
 
 	return rm
 }
 
+func (rm *ReputationModule) AddSource(id string, maddr ma.Multiaddr) error {
+	return rm.sources.Add(source.Source{Id: id, Maddr: maddr})
+}
+
+func (rm *ReputationModule) GetTopMiners(n int) ([]MinerScore, error) {
```

Comment:
For the moment, a method that returns the top N miners with best score.
We could make more rich queries, etc. Limited by imagination.

Reply:
🎉